src/archetype.rs

```diff
@@ -338,6 +339,44 @@ impl Drop for Archetype {
     }
 }
 
+/// A hasher optimized for hashing a single TypeId.
+///
+/// TypeId is already thoroughly hashed, so there's no reason to hash it again.
+/// Just leave the bits unchanged.
+#[derive(Default)]
+pub(crate) struct TypeIdHasher {
+    hash: u64,
+}
+
+impl Hasher for TypeIdHasher {
+    fn write_u64(&mut self, n: u64) {
+        // Only a single value can be hashed, so the old hash should be zero.
+        debug_assert_eq!(self.hash, 0);
+        self.hash = n;
+    }
+
+    // Tolerate TypeId being either u64 or u128.
+    fn write_u128(&mut self, n: u128) {
+        debug_assert_eq!(self.hash, 0);
+        self.hash = n as u64;
+    }
+
+    fn write(&mut self, _bytes: &[u8]) {
+        panic!("`write` should not be called when hashing a TypeId");
```

Comment:
Because `write` is the fallback impl, to reduce the odds of existing code breaking on unforseen future rust versions, I'd prefer to do something conservative here, e.g. take the first 8 bytes, zero-extending.

Reply:
I added the requested paranoia code to handle TypeId being neither u64 nor u128.